### ISSUE TITLE
what the meaning of the method of forward and backward?

### ISSUE THREAD

**Issue by zengjie617789** [keras-rl/keras-rl#246]
could anyone who can tell me what the use of the method of forward and backward?
thank you in advance~

**Comment by mirraaj** [comment]
Hello @zengjie617789! The forward method predicts the action to be taken for any given observation from the given algorithm's policy.

The backward method trains the function approximator (neural network) for a given sample. It also stores the experience replay memory (DQN) for generating samples at the time of learning.

**Comment by zengjie617789** [comment]
thank you very much ; i feel sorry that i did not read the code carefully;
and i have some doubts that what is the mean of nb_steps_warmsup and the metrics?
does it  mean only the number of steps  exceeds the parameter then the neural network would learn and update the weight?
but if it doesnot, what about the trainning?

**Comment by mirraaj** [comment]
You are correct. DQN stores the experience in the memory buffer for the first nb_steps_warmup. This is done to get the required size of batch during experience replay.

Metrics stores the logs of data generated per update. It may be your loss function, reward, output of neural netwok etc. You may want to have a look at https://keras.io/metrics/.

**Comment by zengjie617789** [comment]
thank you@mirraaj; here is another question , i can not get the idea of the graph in tensorboard ; why is there two neural network？ and there are no connections between them ; and as well as the two loss;
i know the theory of DQN; but i can not understand this graph ,can you help me ? thank you
here is the shoot below:
![image](https://user-images.githubusercontent.com/25050291/44952563-d8d53d80-aeb4-11e8-9080-a9d984a3dde6.png)

**Comment by mirraaj** [comment]
Please refer to the following.

https://github.com/keras-rl/keras-rl/blob/ee3614a4083094aa540cfd4b43aaf2f4a4db11fa/rl/agents/dqn.py#L196

The first neural network is the network which you define as your function approximator. The second neural network is your trainable network to compute mean square error loss. Please let me know if I fail to explain anything.

**Comment by zengjie617789** [comment]
@mirraaj  thank you for your patience . I have some questions about the results and i donot know why;
i first trained the agent and get the result graph below:
![image](https://user-images.githubusercontent.com/25050291/44963481-6704fe80-af5d-11e8-87ac-3367e920f5b8.png)
but for some reasons and it is interrupted, so i trained it again using  the trained weight ,but i get the result below:
![image](https://user-images.githubusercontent.com/25050291/44985603-7adb4f80-afb3-11e8-92a0-7154d40fdb75.png)

i donot know what the reason for it ,do you know why ? 
thank you in advance
